src/filters/list.js

```diff
@@ -230,3 +230,96 @@ function isNewListNeeded( previousItem, currentItem ) {
 	// Even with the same id the list does not have to be continuous (#43).
 	return !previousSibling.is( 'ul' ) && !previousSibling.is( 'ol' );
 }
+
+// Paste from Google Docs
+/**
+ * Removes paragraph wrapping content inside list item.
+ *
+ * @param {module:engine/view/element~Element|module:engine/view/documentfragment~DocumentFragment} elementOrDocumentFragment
+ * @param {module:engine/view/upcastwriter~UpcastWriter} writer
+ */
+export function unwrapParagraphInListItem( elementOrDocumentFragment, writer ) {
+	for ( const value of writer.createRangeIn( elementOrDocumentFragment ) ) {
+		const element = value.item;
+
+		if ( element.is( 'li' ) ) {
+			const firstChild = element.getChild( 0 );
+
+			if ( firstChild.is( 'p' ) ) {
+				unwrapSingleElement( firstChild, writer );
+			}
+		}
+	}
+}
+
+/**
+ * Moves nested list inside previous sibling element, what is a proper HTML standard.
+ * There are 2 situations which are fixed in the for loop:
+ *
+ * 1. Move list to previous list item:
+ *
+ *		before                            after:
+ *		OL                                OL
+ *		|-> LI                            |-> LI
+ *		|-> OL                                |-> OL
+ *		    |-> LI                                |-> LI
+ *
+ * 2. Unwrap nested list to avoid situation that UL or OL is direct child of another UL or OL.
+ *
+ *		before                            after:
+ *		OL                                OL
+ *		|-> LI                            |-> LI
+ *		    |-> OL                            |-> OL
+ *		        |-> OL                            |-> LI
+ *		        |   |-> OL                        |-> LI
+ *		        |       |-> OL
+ *		        |           |-> LI
+ *		        |-> LI
+ *
+ *		before                            after:
+ *		OL                                OL
+ *		|-> OL                            |-> LI
+ *		    |-> OL
+ *		         |-> OL
+ *		             |-> LI
+ *
+ * @param {module:engine/view/element~Element|module:engine/view/documentfragment~DocumentFragment} elementOrDocumentFragment
+ * @param {module:engine/view/upcastwriter~UpcastWriter} writer
+ */
+export function fixListIndentation( elementOrDocumentFragment, writer ) {
+	for ( const value of writer.createRangeIn( elementOrDocumentFragment ) ) {
+		const element = value.item;
+
+		// 1. Move nested list as child of sibling list item.
+		if ( element.is( 'li' ) ) {
+			const next = element.nextSibling;
+
+			if ( next && isList( next ) ) {
+				writer.remove( next );
+				writer.insertChild( element.childCount, next, element );
+			}
+		}
+
+		// 2. Unwrap nested list
+		if ( isList( element ) ) {
+			let firstChild = element.getChild( 0 );
+
+			while ( isList( firstChild ) ) {
+				unwrapSingleElement( firstChild, writer );
+				firstChild = element.getChild( 0 );
+			}
+		}
+	}
+}
+
+function isList( element ) {
+	return element.is( 'ol' ) || element.is( 'ul' );
+}
+
+function unwrapSingleElement( element, writer ) {
```

Comment:
This can be moved to the `UpcastWriter` as `unwrapElement` ( the single part is redundant here) to make the name different than `DowncastWriter.unwarp()` which unwraps attribute element.

Reply:
Here it is: ckeditor/ckeditor5-engine#1772